README.md
# GRHttp - A native Ruby Generic HTTP and WebSocket server.
[![Gem Version](https://badge.fury.io/rb/grhttp.svg)](http://badge.fury.io/rb/grhttp)
[![Inline docs](http://inch-ci.org/github/boazsegev/GRHttp.svg?branch=master)](http://www.rubydoc.info/github/boazsegev/GRHttp/master)

This is a native Ruby HTTP and WebSocket multi-threaded server that uses the [GReactor](https://github.com/boazsegev/GReactor) library.

This means it's all ruby, no C or Java code. The code is thread-safe and also supports GReactor's forking... although it might effect your code (especially with regards to websocket broadcasting).

GRHttp is Rack compatible... although [GRHttp is meant to be a step forward](docs/HTTP.md), taking what we learned from our experience using Rack and designing the next generation of servers, as suggested by [José Valim](http://blog.plataformatec.com.br/2012/06/why-your-web-framework-should-not-adopt-rack-api/).

## How do I get it?

Add this line to your application's Gemfile:

```ruby
gem 'grhttp'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ [sudo] gem install grhttp


## To Rack or Not to Rack?

### We _can_...

It's true that we _can_ comfortably use GRHttp as a Rack server - just enter the folder for your favorit Rack compatible application (such as Rails or Sinatra) and write (remember to edit your `Gemfile` first):

    $ rackup -s grhttp -p 3000

For most frameworks (Rails/Sinatra), you can also add the `grhttp` gem to your gemfile and it will replace WEBrick as the default fallback server.

### We _should?_

So, we know we _can_, but _should_ we?

There is no simple answer. Using GRHttp through Rack is a bit slower by nature and uses an older (and heavier) server workflow design. on the other hand, if you have an existing application, what is the sense of re-writing all your code?

### We don't have to decide?!

The wonderful thing is, we really don't have to decide - we can mix and match GRHttp handlers with the Rack app (native handlers have priority).

Here is a quick example - place this `config.ru` file in a new folder and see for yourself:

```ruby
# config.ru
ENV['RACK_ENV'] = 'production'

require 'grhttp'

GRHttp.listen do |request, response|
  if request[:path] == '/grhttp'
    response << "Hello from GRHttp!"
    true
  else
    false
  end
end

rack_app = Proc.new do |env|
  [200, {"Content-Type" => "text/html", "Content-Length" => '16' }, ['Hello from Rack!'] ]
end

run rack_app
```

Run this code using `$ rackup -s grhttp -p 3000` (remember to be in the folder with the `config.ru` file)... Now test:

* GRHttp native server on: [http:localhost:3000/grhttp](http:localhost:3000/grhttp)
* GRHttp-Rack server on: [http:localhost:3000/rack](http:localhost:3000/rack)

This is How the [Plezi framework](https://github.com/boazsegev/plezi), which uses GRHttp's native server allows you to both run the Plezi framework's app and your Rails/Sinatra app at the same time - letting you use websockets and advanced features while still maintaining your existing codebase.

### How does it stack up against other servers?

The greatest sin is believing 'Hello World' apps are a good benchmark - they are NOT. For instance, replacing my Thin server on a production app with GRHttp had no meaningful and added Websockets to the mix... In the end of the day, the server is usually NOT the bottleneck.

But since "Hello World" apps the common way to isolate the server component, I tried doing some benchmarking and was fairly happy with the results.

But... Since every update might change the numbers, I removed the benchmarks from this README file. You can run the benchmarks on your own using the following code and see for yourself:

This was the Rackapp tested:

```ruby
# config.ru
ENV['RACK_ENV'] = 'production'
app = Proc.new do |env|
  [200, {"Content-Type" => "text/html", "Content-Length" => '16' }, ['Hello from Rack!'] ]
end
run app
```

The native GRHttp app tested was a terminal command:

      $ ruby -rgrhttp -e "GRHttp.start { GRHttp.listen {'Hello from GRHttp :-)'} }"

The hibrid-app tested was the code in the previous section.

Benchmarks were executed using `wrk` since not all servers answered `ab` (the issue is probably due to `close` vs. `keep-alive` connections over HTTP/1 while `wrk` uses HTTP/1.1):

     $ wrk -c400 -d10 -t12 http://localhost:3000/

\*In contrast to other Rack servers, GRHttp parses all of the HTTP request, offering you a read/write cookie-jar and parsing both POST data and query string data. The smart `request` and `response` is available to native apps.

## Usage

GRHttp allows you to write a quick web service based of the [GReactor](https://github.com/boazsegev/GReactor) library - So you get Asynchronous events, timed event and a Server (or servers) - all in a native Ruby package.

You can use GRHttp for both HTTP and Websocket services and even write a full blown framework, such as the [Plezi HTTP and Websocket WebApp Framework](https://github.com/boazsegev/plezi) that uses GRHttp as it's server of choice.

### HTTP Server

Allow me to welcome the world famous 'Hello World'!

```ruby
require 'grhttp'
GRHttp.start {   GRHttp.listen { 'Hello World!' }      }
exit # exit between examples, to clear the listening services.
```

* [Test it at http://localhost:3000](http://localhost:3000).

As you may have noticed, by returning a string, the server automatically appended the string to the end of the response. This might be limited, so there's a better way to do this.

Also, we don't have to be hanging around while the server works... we can keep running tasks with the server in the background:

```ruby
require 'grhttp'

GRHttp.start

GRHttp.listen(timeout: 3, port: 3000) do |request, response|
   response.cookies[:name] = 'setting cookies is easy'
   response.flash[Time.now.to_s] = 'this cookie will selfdestruct on your next visit.'
   response << 'Hello!'
end

GRHttp.on_shutdown { GRHttp.clear_listeners; }

puts "We can keep working... try opening a page, go on... I'll wait..."
sleep 5

puts 'We can even restart the server:'
GRHttp.restart # restart doesn't invoke the shutdown callbacks.

puts 'Now we'll hang, press ^C to stop the service.'
GRHttp.join

```

* [Test it at http://localhost:3000](http://localhost:3000).
* [Test The SSL version at http://localhost:3030](http://localhost:3030).

Ammm... This example was too quick to explain much or show off all the bells and whistles, so, let's try again, this time - Bigger, better, and... websockets, anyone?

### Websocket services

WebSockets are also supported. Here's a "Hello World!" variation with a WebSocket Echo Server that both Boradcasts and Unicasts data to different websockets (semi-chatroom).

We'll be using the same handler to handle regular HTTP requests, upgrade requests and WebSocket requests, all at once. This might better simulate an HTTP Router situation, which routes all types of requests to their corrects controller (or controller class):

```ruby
module MyServer
    module_function

    # handles HTTP

    def call request, response
      if request.upgrade? # upgrade to websockets?
        return false if request.path == '/refuse'
        return self
      end
      return false if request.path == '/refuse'
      response << "Hello #{'SSL ' if request.ssl?}World!\r\n\r\n"
      response << "To check the websocket server use: http://www.websocket.org/echo.html\r\n\r\n"
      response << "The Request:\r\n#{request.to_s}"
    end

    # handles WebSockets

    def on_open ws
      puts 'WebSocket Open!'
      ws << 'System: Welcome!'
      ws.autopong # sets auto-pinging, to keep alive

      @first ||= ws.uuid # save the first connection's uuid.
      if ws.uuid == @first
        ws << "System: You're the first one here, so you get special notifications :-)"
      else
        ws.unicast @first, "JOINED: Someone just joined the chat."
      end
    end
    def on_message ws
      ws << "You: #{ws.data}"
      ws.broadcast "Someone: #{ws.data}"
    end

    def on_broadcast ws
      ws << ws.data
      true
    end

    def on_close ws
      ws.unicast @first, "LEFT: Someone just left the chat."
      puts 'WebSocket Closed!'
    end
end

GRHttp.start do
    GRHttp.listen upgrade_handler: MyServer, http_handler: MyServer
    GRHttp.listen port: 3030, ssl: true, upgrade_handler: MyServer, http_handler: MyServer
    GRHttp.on_shutdown { GRHttp.clear_listeners;  GRHttp.info 'Clear :-)'}
    GRHttp.on_shutdown { puts 'Shutdown and cleanup complete.'}
end
```

* [Test the `/refuse` path at http://localhost:3000/refuse](http://localhost:3000/refuse).
* [Test the wesocket echo server using http://websocket.org/echo.html](http://websocket.org/echo.html).


## Websocket Client

GRHttp can also act as a client for either a semi-synchronous or asynchronous websocket connections.

GRHttp's websocket client was designed for simplicity, as it started it's life as a testing tool.

Here is a very simple implementation that leverages the fact that `GRHttp.ws_connect` will hang until a connection is established - making the code very simple to write:

```ruby
require 'grhttp'

client = GRHttp.ws_connect 'ws://echo.websocket.org' do |ws|
    puts "Received >> #{ws.data}"
    ws.close if ws.data =~ /^(bye|quit|exit|close)/i
end

client << "Echo this!"

client << "Bye" # => true

client.closed?

client << 'test' # => false

```

The block passed acted as our `on_message` callback. But what about an `on_close` callback, or an `on_connect` one... and what about some cookies?

Also, what if we don't want our code to block? Or maybe we want to connect to an SSL connection?

Okay, Okay... So many questions. No worries, here's a more complete example:


```ruby
require 'grhttp'

# Start GRHttp, if it isn't running in the background already.
# We're about to use asynchronous code.
number_of_threads = 30
GRHttp.start number_of_threads

# the following options will be used to setup the connection
options = {}

# The on_close callback
options[:on_close] = Proc.new {|ws| puts 'Connection closed.' }

# The on_open callback
options[:on_open] = Proc.new do |ws|
    # notify that the connection is open
    puts 'Connection opened.'
    # use the ws.ssl? method to make sure it's SSL
    puts 'This is an SSL connection' if ws.ssl?
    # write some stuff
    ws << 'Hello World!'
    10.times {|i|  ws << "Countdown: #{10-i}\n"}
    # say goodbye
    ws << 'Bye Bye'
end

# The on_message callback (instead of using a block like before)
options[:on_message] = Proc.new do |ws|
    # print out any incoming data
    puts "Received >> #{ws.data}"
    # close if we received a goodbye keyword
    ws.close if ws.data =~ /^(bye|quit|exit|close|goodbye)/i
end

# Cookies anyone?
options[:cookies] = {my_cookie_name: 'my cookie value'}

# Run the connection asynchronously
GRHttp.run_async { GRHttp.ws_connect 'wss://echo.websocket.org', options }

# Block, so we can see our code execute.
puts "Press ^C to stop the server"
GRHttp.join
```

Easy.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake false` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/boazsegev/grhttp.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
#Change Log

***

Change log v.0.1.0

**Update**: Update the GReactor version to version 0.1.0, adopting it's new IO API.

**API Changes**: The structure and API of the GRHttp server is now somewhat different. Ease of use had been favored over control in some cases.

**Update**: Updated the autoping feature, as well as some of the logging information supplied. retouched the Rack support, reviewed the FileSession initialization (it's faster now), prepared some groundwork for websocket extentions... all sorts of goodies under the hood.

**Feature**: The Websocket extention API is now complete and it is now possible to write Websocket extentions sch as the `permessage-deflate` extention.

**Feature**: The Websocket server now supports the `permessage-deflate` extention:

*   The extention will compress UTF-8 text data (but not binary data) longer than 16 bytes.
*   The extention is supported by default and can be disabled (unregistered) by calling: `GRHttp.allow_websocket_deflate false`
*   The extention was tested on Chrome, Safari and Firefox, but it will only work on Chrome while the extention will be automatically ignored when connecting with other browsers (Safari doesn't use `permessage-deflate` and Firefox has an issue that causes compression to fail, so it's ignored).
* The extention wasn't tested with IE nor Edge.

**Fix**: Fixed an issue with the websocket connect shortcut method (`GRHttp.ws_connect` vs. `GRHttp::WSClient.connect`), which caused options passed to the shortcut to be ignored. This is now fixed.

**Deprecation notice**: `request[:scheme]` or `request.scheme` should be used instead of `request[:requested_protocol]` or `request.protocol`.

***

Change log v.0.0.23

**Update**: Update the required GReactor version.

**Fix**: fixed cookie name requirements for the WSClient.

**API**: added  `ws_connect` API for quick access to the WSClient.connect.

***

Change log v.0.0.22

**Feature**: General Websocket client (GRHttp::WSClient) improvements. i.e., added custom header options for basic cookie support and HTTP variables/authentication for the websocket client's handshake.

**Cookie errors**: GRHttp will now raise an error if an illegal cookie name is attempted.

***

Change log v.0.0.21

**Fix**: Fixed an issue with Firefox's implementation of Websockets, which sends a connection string that includes both the `keep-alive` and the `upgrade` directives (the standard is to only send the `upgrade` directive).

***

Change log v.0.0.20

**Session support**: Basic serverside session support for memory stored session data is now integrated into the response object (which sets the session ID). It's easy to replace the memory session storage with a session storage object that supports DB/File storage by providing a compatible Session Storage object that andwers to `fetch(id)` and returns a self managing Hash like Session Object.

**WS Client** Added the `closed?` method, to check if the websocket connection is still open.

**Fix**: Cookie values were saved in the cookie-jar with the newline marker. This is now fixed by striping the newline marker between HTTP all headers prior to processing.

**Fix**: minor HTTP parsing fixes for performance enhancement introduced in v.0.0.19.

***

Change log v.0.0.19

**Performance**: Reviewd the code for the params Hash propagation and slightly improved the performance.

**Performance**: Reviewd the code for Rack support and delegated more control to Rack (this can also be bad, but it is what it is).

**Settings** Including GRHttp in your gemfile will now automatically replace the Webrick server as the default for Rack. 
**Deprecation**: For performance reasons, Rack apps will not recieve the writable cookie-jar nor the parsed params priorly available (`gr.cookies` and `gr.params` were not utilized by past working Rack apps and Rack support is meant to help run new code in parallel to existing apps).

**Fix**: Fixed an issue with the response's logging, which prevented the response time (in miliseconds) from printing out response times less then a milisecond.

***

Change log v.0.0.18

**Fix**: Fixed an issue with the method used to extract data into the params (or cookie) Hash. The issue caused an exception to be raised but the correct value to be placed in the Hash.

***

Change log v.0.0.17

**Fix**: Fixed a typo in the GRHttp's Rack variable naming (was `pl.cookies` instead of `gr.cookies`, same issue with `gr.params`).

**Update**: better Rack support.

***

Change log v.0.0.16

**Fix**: GRHttp will better respect requests from HTTP/1 or with the header 'Connection: close'.

***

Change log v.0.0.15

**Wow**: This is GRHttp's first version as a Dual-Rack-GRHttp server - this means you can use GRHttp both as a Rack server (with all of Rack's limitations) while using GRHttp's native features outside of Rack at the same time (see the [README]{README.md}) for more details.

***

Change log v.0.0.14

**update**: attempts to encode incoming data as UTF8 when the encoding conforms.

***

Change log v.0.0.13

**performance**: clears HTTP and Websocket parsing (and logging) memory faster.

**update**: requires the updated GReactor (v.0.0.11)

***

Change log v.0.0.12

**update**: requires the updated GReactor (v.0.0.10)

***

Change log v.0.0.11

**Fix**: Fixed an issue where websockets might be answerwed if the response was sent and the update_handler return true (using the `response.finish`).

***

Change log v.0.0.10

**Fix**: Fixed an issue where the last byte of an incoming Websocket message wouldn't be unmasked - resulting in corrupt data.

***

Change log v.0.0.9

Changelog will start loging after the first release that will be viable for production testing.

For now, this version introduces great performance upgrade and some minor bug fixes for the Websocket protocol handling.

lib/grhttp/hpack.rb
#!/usr/bin/env ruby

module GRHttp
	class HTTP2 < GReactor::Protocol
		class HPACK
			class IndexTable
				attr_reader :size
				attr_accessor :max_size
				def initialize
					@list = []
					@max_size = @size = 4_096 # initial defaul size by standard
					@actual_size = 0
				end
				def [] index
					raise "HPACK Error - invalid header index: 0" if index == 0
					return STATIC_LIST[index] if index < STATIC_LENGTH
					raise "HPACK Error - invalid header index: 0" if @list.count <= (index - STATIC_LENGTH)
					@list[index - STATIC_LENGTH]
				end
				alias :get_index :[]
				def get_name index
					get_index(index)[0]
				end
				def insert *field
					@list.unshift field
					field.each {|f| @actual_size += f.to_s.bytesize}; @actual_size += 32
					resize
					field
				end
				def find *field
					index = STATIC_LIST.index(field)
					index ||= @list.index(feild)
					index ? (index + STATIC_LENGTH) : nil
				end
				def find_name name
					index = 1
					while STATIC_LIST[index]
						return index if STATIC_LIST[index][0] == name
						index += 1
					end
					index = 0
					while @list[index]
						return index+STATIC_LENGTH if @list[index][0] == name
						index += 1
					end
					nil
				end
				def resize value = nil
					@size = value if value && value < @max_size
					while (@actual_size > @size) && @list.any?
						@list.pop.each {|i| @actual_size -= i.to_s.bytesize}
						@actual_size -= 32
					end
					self
				end
			end

			def initialize
				@decoding_list = IndexTable.new
				@encoding_list = IndexTable.new
			end

			def decode data
				data = StringIO.new data
				results = {}
				while (field = decode_field(data))
					results[field[0]] ? (results[field[0]].is_a?(String) ? (results[field[0]] = [results[field[0]], field[1]]) : (results[field[0]] << field[1]) ) : (results[field[0]] = field[1]) if field[1]
				end
				results
			end
			def encode headers = {}
				buffer = ''
				headers.each {|k, v| buffer << encode_field(b,v) if v}
				buffer
			end

			protected
			def decode_field data # expects a StringIO or other IO object
				byte = data.getbyte
				return nil unless byte
				if byte[7] == 1 # 0b1000_0000 == 0b1000_0000
					# An indexed header field starts with the '1' 1-bit pattern, followed by the index of the matching header field, represented as an integer with a 7-bit prefix (see Section 5.1).
					num = extract_number data, byte, 1
					@decoding_list[num]
				elsif byte & 192 == 64 # 0b1100_0000 == 0b0100_0000
					# A literal header field with incremental indexing representation starts with the '01' 2-bit pattern.
					# If the header field name matches the header field name of an entry stored in the static table or the dynamic table, the header field name can be represented using the index of that entry. In this case, the index of the entry is represented as an integer with a 6-bit prefix (see Section 5.1). This value is always non-zero.
					# Otherwise, the header field name is represented as a string literal (see Section 5.2). A value 0 is used in place of the 6-bit index, followed by the header field name.
					num = extract_number data, byte, 2
					field_name = (num == 0) ? extract_string(data) : @decoding_list.get_name(num)
					field_value = extract_string(data)
					@decoding_list.insert field_name, field_value
				elsif byte & 224 # 0b1110_0000 == 0
					# A literal header field without indexing representation starts with the '0000' 4-bit pattern.
					# If the header field name matches the header field name of an entry stored in the static table or the dynamic table, the header field name can be represented using the index of that entry.
					# In this case, the index of the entry is represented as an integer with a 4-bit prefix (see Section 5.1). This value is always non-zero.
					# Otherwise, the header field name is represented as a string literal (see Section 5.2) and a value 0 is used in place of the 4-bit index, followed by the header field name.
					# OR
					# A literal header field never-indexed representation starts with the '0001' 4-bit pattern + 4+ bits for index
					num = extract_number data, byte, 4
					field_name = (num == 0) ? extract_string(data) : @decoding_list.get_name(num)
					field_value = extract_string(data)
					[field_name, field_value]
				elsif byte & 224 == 32 # 0b1110_0000 == 0b0010_0000
					# A dynamic table size update starts with the '001' 3-bit pattern
					# followed by the new maximum size, represented as an integer with a 5-bit prefix (see Section 5.1).
					@decoding_list.resize extract_number(data, byte, 5)
					[].freeze
				else
					raise "HPACK Error - invalid field indicator."
				end
			end
			def encode_field name, value
				if value.is_a?(Array)
					return (value.map {|v| encode_field name, v} .join)
				end
				if name == 'set-cookie'
					buffer = ''
					buffer << pack_number( 55, 16, 4)
					buffer << pack_string(value)
					return buffer
				end
				index = @encoding_list.find(name, value)
				return pack_number( index, 1, 1) if index
				index = @encoding_list.find_name name
				@encoding_list.insert name, value
				buffer = ''
				if index
					buffer << pack_number( index, 64, 2)
				else
					buffer << pack_number( 0, 64, 2)
					buffer << pack_string(name)
				end
				buffer << pack_string(value)
				buffer
			end
			def extract_number data, prefix, prefix_length
				mask = 255 >> prefix_length
				return prefix & mask unless (prefix & mask) == mask
				count = prefix = 0
				loop do
					c = data.getbyte
					prefix = prefix | ((c & 127) << (7*count))
					break if c[7] == 0
					count += 1
				end
				prefix + mask
			# rescue e =>
			# 	raise "HPACK Error - number input invalid"
			end
			def pack_number number, prefix, prefix_length
				n_length = 8-prefix_length
				if (number + 1 ).bit_length <= n_length
					return ((prefix << n_length) | number).chr
				end
				prefix = [(prefix << n_length) | (2**n_length - 1)]
				number -= 2**n_length - 1
				loop do
					prefix << ((number & 127) | 128)
					number = number >> 7
					break if number == 0
				end
				(prefix << (prefix.pop & 127)).pack('C*'.freeze)
			end
			def pack_string string, deflate = true
				string = deflate(string) if deflate
				(pack_number(string.bytesize, (deflate ? 1 : 0), 1) + string).force_encoding ::Encoding::ASCII_8BIT
			end
			def extract_string data
				byte = data.getbyte
				hoffman = byte[7] == 1
				length = extract_number data, byte, 1
				if hoffman
					inflate data.read(length)
				else
					data.read length
				end
			end
			def inflate data
				data = StringIO.new data
				str = ''
				buffer = ''
				until data.eof?
					byte = data.getbyte
					8.times do |i|
						buffer << byte[7-i].to_s
						if HUFFMAN[buffer]
							str << HUFFMAN[buffer].chr rescue raise("HPACK Error - Huffman EOS found")
							buffer.clear
						end
					end
				end
				raise "HPACK Error - Huffman padding too long (#{buffer.length}): #{buffer}" if buffer.length > 29
				str
			end
			def deflate data
				str = ''
				buffer = ''
				data.bytes.each do |i|
					buffer << HUFFMAN.key(i)
					if (buffer % 8) == 0
						str << [buffer].pack('b*')
						buffer.clear
					end
				end
				(8-(buffer.bytesize % 8)).times { buffer << '1'}
				str << [buffer].pack('b*')
				buffer.clear
				str
			end
			STATIC_LIST = [ nil,
				[:authority],
				[:method, "GET" ],
				[:method, "POST" ],
				[:path, "/" ],
				[:path, "/index.html" ],
				[:scheme, "http" ],
				[:scheme, "https" ],
				[:status, "200" ],
				[:status, "204" ],
				[:status, "206" ],
				[:status, "304" ],
				[:status, "400" ],
				[:status, "404" ],
				[:status, "500" ],
				["accept-charset"],
				["accept-encoding", "gzip, deflate" ],
				["accept-language"],
				["accept-ranges"],
				["accept"],
				["access-control-allow-origin"],
				["age"],
				["allow"],
				["authorization"],
				["cache-control"],
				["content-disposition"],
				["content-encoding"],
				["content-language"],
				["content-length"],
				["content-location"],
				["content-range"],
				["content-type"],
				["cookie"],
				["date"],
				["etag"],
				["expect"],
				["expires"],
				["from"],
				["host"],
				["if-match"],
				["if-modified-since"],
				["if-none-match"],
				["if-range"],
				["if-unmodified-since"],
				["last-modified"],
				["link"],
				["location"],
				["max-forwards"],
				["proxy-authenticate"],
				["proxy-authorization"],
				["range"],
				["referer"],
				["refresh"],
				["retry-after"],
				["server"],
				["set-cookie"],
				["strict-transport-security"],
				["transfer-encoding"],
				["user-agent"],
				["vary"],
				["via"],
				["www-authenticate"] ].map! {|a| a.map! {|s| s.is_a?(String) ? s.freeze : s } && a.freeze if a}
			STATIC_LENGTH = STATIC_LIST.length

			HUFFMAN = [
				"1111111111000", 
				"11111111111111111011000", 
				"1111111111111111111111100010", 
				"1111111111111111111111100011", 
				"1111111111111111111111100100", 
				"1111111111111111111111100101", 
				"1111111111111111111111100110", 
				"1111111111111111111111100111", 
				"1111111111111111111111101000", 
				"111111111111111111101010", 
				"111111111111111111111111111100", 
				"1111111111111111111111101001", 
				"1111111111111111111111101010", 
				"111111111111111111111111111101", 
				"1111111111111111111111101011", 
				"1111111111111111111111101100", 
				"1111111111111111111111101101", 
				"1111111111111111111111101110", 
				"1111111111111111111111101111", 
				"1111111111111111111111110000", 
				"1111111111111111111111110001", 
				"1111111111111111111111110010", 
				"111111111111111111111111111110", 
				"1111111111111111111111110011", 
				"1111111111111111111111110100", 
				"1111111111111111111111110101", 
				"1111111111111111111111110110", 
				"1111111111111111111111110111", 
				"1111111111111111111111111000", 
				"1111111111111111111111111001", 
				"1111111111111111111111111010", 
				"1111111111111111111111111011", 
				"010100", 
				"1111111000", 
				"1111111001", 
				"111111111010", 
				"1111111111001", 
				"010101", 
				"11111000", 
				"11111111010", 
				"1111111010", 
				"1111111011", 
				"11111001", 
				"11111111011", 
				"11111010", 
				"010110", 
				"010111", 
				"011000", 
				"00000", 
				"00001", 
				"00010", 
				"011001", 
				"011010", 
				"011011", 
				"011100", 
				"011101", 
				"011110", 
				"011111", 
				"1011100", 
				"11111011", 
				"111111111111100", 
				"100000", 
				"111111111011", 
				"1111111100", 
				"1111111111010", 
				"100001", 
				"1011101", 
				"1011110", 
				"1011111", 
				"1100000", 
				"1100001", 
				"1100010", 
				"1100011", 
				"1100100", 
				"1100101", 
				"1100110", 
				"1100111", 
				"1101000", 
				"1101001", 
				"1101010", 
				"1101011", 
				"1101100", 
				"1101101", 
				"1101110", 
				"1101111", 
				"1110000", 
				"1110001", 
				"1110010", 
				"11111100", 
				"1110011", 
				"11111101", 
				"1111111111011", 
				"1111111111111110000", 
				"1111111111100", 
				"11111111111100", 
				"100010", 
				"111111111111101", 
				"00011", 
				"100011", 
				"00100", 
				"100100", 
				"00101", 
				"100101", 
				"100110", 
				"100111", 
				"00110", 
				"1110100", 
				"1110101", 
				"101000", 
				"101001", 
				"101010", 
				"00111", 
				"101011", 
				"1110110", 
				"101100", 
				"01000", 
				"01001", 
				"101101", 
				"1110111", 
				"1111000", 
				"1111001", 
				"1111010", 
				"1111011", 
				"111111111111110", 
				"'", 
				"11111111111101", 
				"1111111111101", 
				"1111111111111111111111111100", 
				"11111111111111100110", 
				"1111111111111111010010", 
				"11111111111111100111", 
				"11111111111111101000", 
				"1111111111111111010011", 
				"1111111111111111010100", 
				"1111111111111111010101", 
				"11111111111111111011001", 
				"1111111111111111010110", 
				"11111111111111111011010", 
				"11111111111111111011011", 
				"11111111111111111011100", 
				"11111111111111111011101", 
				"11111111111111111011110", 
				"111111111111111111101011", 
				"11111111111111111011111", 
				"111111111111111111101100", 
				"111111111111111111101101", 
				"1111111111111111010111", 
				"11111111111111111100000", 
				"111111111111111111101110", 
				"11111111111111111100001", 
				"11111111111111111100010", 
				"11111111111111111100011", 
				"11111111111111111100100", 
				"111111111111111011100", 
				"1111111111111111011000", 
				"11111111111111111100101", 
				"1111111111111111011001", 
				"11111111111111111100110", 
				"11111111111111111100111", 
				"111111111111111111101111", 
				"1111111111111111011010", 
				"111111111111111011101", 
				"11111111111111101001", 
				"1111111111111111011011", 
				"1111111111111111011100", 
				"11111111111111111101000", 
				"11111111111111111101001", 
				"111111111111111011110", 
				"11111111111111111101010", 
				"1111111111111111011101", 
				"1111111111111111011110", 
				"111111111111111111110000", 
				"111111111111111011111", 
				"1111111111111111011111", 
				"11111111111111111101011", 
				"11111111111111111101100", 
				"111111111111111100000", 
				"111111111111111100001", 
				"1111111111111111100000", 
				"111111111111111100010", 
				"11111111111111111101101", 
				"1111111111111111100001", 
				"11111111111111111101110", 
				"11111111111111111101111", 
				"11111111111111101010", 
				"1111111111111111100010", 
				"1111111111111111100011", 
				"1111111111111111100100", 
				"11111111111111111110000", 
				"1111111111111111100101", 
				"1111111111111111100110", 
				"11111111111111111110001", 
				"11111111111111111111100000", 
				"11111111111111111111100001", 
				"11111111111111101011", 
				"1111111111111110001", 
				"1111111111111111100111", 
				"11111111111111111110010", 
				"1111111111111111101000", 
				"1111111111111111111101100", 
				"11111111111111111111100010", 
				"11111111111111111111100011", 
				"11111111111111111111100100", 
				"111111111111111111111011110", 
				"111111111111111111111011111", 
				"11111111111111111111100101", 
				"111111111111111111110001", 
				"1111111111111111111101101", 
				"1111111111111110010", 
				"111111111111111100011", 
				"11111111111111111111100110", 
				"111111111111111111111100000", 
				"111111111111111111111100001", 
				"11111111111111111111100111", 
				"111111111111111111111100010", 
				"111111111111111111110010", 
				"111111111111111100100", 
				"111111111111111100101", 
				"11111111111111111111101000", 
				"11111111111111111111101001", 
				"1111111111111111111111111101", 
				"111111111111111111111100011", 
				"111111111111111111111100100", 
				"111111111111111111111100101", 
				"11111111111111101100", 
				"111111111111111111110011", 
				"11111111111111101101", 
				"111111111111111100110", 
				"1111111111111111101001", 
				"111111111111111100111", 
				"111111111111111101000", 
				"11111111111111111110011", 
				"1111111111111111101010", 
				"1111111111111111101011", 
				"1111111111111111111101110", 
				"1111111111111111111101111", 
				"111111111111111111110100", 
				"111111111111111111110101", 
				"11111111111111111111101010", 
				"11111111111111111110100", 
				"11111111111111111111101011", 
				"111111111111111111111100110", 
				"11111111111111111111101100", 
				"11111111111111111111101101", 
				"111111111111111111111100111", 
				"111111111111111111111101000", 
				"111111111111111111111101001", 
				"111111111111111111111101010", 
				"111111111111111111111101011", 
				"1111111111111111111111111110", 
				"111111111111111111111101100", 
				"111111111111111111111101101", 
				"111111111111111111111101110", 
				"111111111111111111111101111", 
				"111111111111111111111110000", 
				"11111111111111111111101110", 
				"111111111111111111111111111111"].map.with_index {|s, i| [s, i]} .to_h
		end
	end
end